### PR TITLE
dist: ethos: add tcp socket support

### PR DESCRIPTION
**Pull request by cladmi**
Add ethos support to use a tcp socket instead of a TTY.
Use 'tcp:hostname[:port]' as serial argument.

Default port is 20000 which is used in on IoT-LAB.

Known issue: hostname cannot be an ipv6 address, because of the ':'.

### CODE REVIEW

File: dist/tools/ethos/ethos.c [hunk 1]
```diff
@@ -267,6 +277,134 @@ static void _clear_neighbor_cache(const char *ifname)
     }
 }
 
+int parse_tcp_arg(char *name, char **host, char **port)
+{
+    char *host_str;
+    char *port_str;
+
+    // Remove 'tcp:'
```
Comment (aabadie) [inline dist/tools/ethos/ethos.c]: use C style comments /* */. There are others below.

File: dist/tools/ethos/ethos.c [hunk 2]
```diff
@@ -267,6 +277,134 @@ static void _clear_neighbor_cache(const char *ifname)
...
+
+int set_socket_timeout(int sfd)
+{
+    struct timeval timeout = {
+        .tv_sec = 0,
+        .tv_usec = TTY_TIMEOUT_MS * 1000,
+    };
+
+    if (setsockopt(sfd, SOL_SOCKET, SO_RCVTIMEO,
+                (char *)&timeout, sizeof(timeout)) == -1) {
```
Comment (aabadie) [inline dist/tools/ethos/ethos.c]: indentation seems wrong to me, it should be aligned with the ligne above.

File: dist/tools/ethos/ethos.c [hunk 3]
```diff
@@ -62,7 +70,8 @@ int set_serial_attribs (int fd, int speed, int parity)
                                         /* no canonical processing*/
     tty.c_oflag = 0;                    /* no remapping, no delays*/
     tty.c_cc[VMIN]  = 0;                /* read doesn't block*/
-    tty.c_cc[VTIME] = 5;                /* 0.5 seconds read timeout*/
+    tty.c_cc[VTIME] = TTY_TIMEOUT_MS / 100; /* 0.5 seconds read timeout*/
+                                            /* in tenths of a second*/
```
Comment (aabadie) [inline dist/tools/ethos/ethos.c]: tenth

File: dist/tools/ethos/ethos.c [hunk 4]
```diff
@@ -267,6 +277,134 @@ static void _clear_neighbor_cache(const char *ifname)
     }
 }
 
+int parse_tcp_arg(char *name, char **host, char **port)
+{
+    char *host_str;
+    char *port_str;
+
+    // Remove 'tcp:'
+    name = &name[sizeof(TCP_DEV) -1];
```
Comment (aabadie) [inline dist/tools/ethos/ethos.c]: missing space after `-`

File: dist/tools/ethos/ethos.c [hunk 5]
```diff
@@ -267,6 +277,134 @@ static void _clear_neighbor_cache(const char *ifname)
...
+        return -1;
+    }
+
+    set_serial_attribs(serial_fd, B115200, 0);
+    set_blocking(serial_fd, 1);
+
+    return serial_fd;
+}
+
+int open_serial(char *name)
```
Comment (aabadie) [inline dist/tools/ethos/ethos.c]: The function name could be improved:
`_open_connection` here and more specific names for serial and tcp connections (above): `_open_serial_connection` and `_open_tcp_connection`

**Comment by aabadie** [comment]
Awesome work @cladmi :)
Except a few nit-picks in the code, this is good to me. I just need to test it on IoT-LAB before giving it an ACK.

**Comment by aabadie** [comment]
@cladmi, seems that you need to rebase ;)

File: dist/tools/ethos/ethos.c (hunk 3 above)
Comment (cladmi) [inline dist/tools/ethos/ethos.c]: I kept it to 'tenths' because it's a plural

**Comment by cladmi** [comment]
I rebased and took the opportunity to change the arguments.
I used the 'baudrate' additional argument to pass a possible 'port'.

So now the usage is `./ethos <tap> tcp:host [port]`

Giving an ipv6 address as host is now possible and is working.

This argument parsing changed for both serial and tcp and needs reviewing.
I am also not so proud of the  `usage` function. I did not know how to write it.
If I needed to change the basic line and make it more complex, or if I need to add other help lines to describe each argument. Concision versus verbosity.

File: dist/tools/ethos/ethos.c (hunk 3 above)
Comment (aabadie) [inline dist/tools/ethos/ethos.c]: google agrees with you.

**Comment by cladmi** [comment]
Usages ideas:

```
Usage: ethos <tap> <serial> [baudrate] |
             <tap> tcp:<host> [port]
```

```
Usage: ethos <tap> < <serial> [baudrate] | tcp:<host> [port] >
```

```
Usage: ethos <tap> < <serial> [baudrate] | tcp:<host> [port] >
```

```
Usage: ethos <tap> <serial | tcp:<host>> [baudrate|port]
```

**Comment by aabadie** [comment]
The new command line argument change is a valuable improvement. I would go for verbosity in the usage message, something like this:
```
$ ./ethos
usage: ethos <tap> <resource> [baudrate] [port]
   <resource> can be either a tty or a hostname (with format tcp:<hostname>)
   [baudrate] is only uses when <resource> is a tty
   [port] is only used when <resource> is a hostname
```

**Comment by aabadie** [comment]
@OlegHahm, @kaspar030 any suggestions ?

**Comment by aabadie** [comment]
Once we find a consensus about the usage message, I'll give an ACK.

**Comment by aabadie** [comment]
@cladmi, I would use this usage: 
```
Usage: ethos <tap> < <serial> [baudrate] | tcp:<host> [port] >
```
I find it clearer and less verbose.

Can you apply the change ? and then rebase ?

**Comment by kaspar030** [comment]
@aabadie I find that hard to parse...

What about:

    Usage: ethos <tap> <serial> [baudrate]
           ethos <tap> tcp:<host> [port]

**Comment by aabadie** [comment]
ping @cladmi 
I'm fine with @kaspar030 's suggestion.

**Comment by cladmi** [comment]
I updated the usage and rebased.

**Comment by aabadie** [comment]
@cladmi , thanks ! Can you squash and keep only the first commit ? Then I'll give an ACK

**Comment by cladmi** [comment]
Done

**Comment by aabadie** [comment]
Let's now wait for Murdock.

**Comment by aabadie** [comment]
All green => ACK

I added this one for the next release (2016.10), so it will be merged after 2016.07.
